pkg/capacity/provision.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package capacity
+
+import (
+	"context"
+
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/sig-storage-lib-external-provisioner/v6/controller"
+)
+
+type provisionWrapper struct {
+	controller.Provisioner
+	c *Controller
+}
+
+func NewProvisionWrapper(p controller.Provisioner, c *Controller) controller.Provisioner {
+	return provisionWrapper{
+		Provisioner: p,
+		c:           c,
+	}
+}
+
+func (p provisionWrapper) Provision(ctx context.Context, options controller.ProvisionOptions) (pv *v1.PersistentVolume, state controller.ProvisioningState, err error) {
```

Comment:
Because none of the two reasons for using a pointer receiver in https://tour.golang.org/methods/8 apply: the methods don't need to modify the instance and the instance is small (interface + pointer).

I know that it is more complex (https://stackoverflow.com/questions/27775376/value-receiver-vs-pointer-receiver), but value receivers seems appropriate here.

Reply:
The golang code review wiki (https://github.com/golang/go/wiki/CodeReviewComments#receiver-type) recommends:

> If the receiver is a struct, array or slice and any of its elements is a pointer to something that might be mutating, prefer a pointer receiver, as it will make the intention more clear to the reader.

But this is totally a nitpick. Either option lgtm

Reply:
That is indeed more precise than the initial introduction and the StackOverflow discussion. However, the intention here isn't to allow mutation. So in that sense using a pointer would *obscure* the intention instead of making it clear. Or am I missing something? I'd be happy to learn more about Go here :sweat_smile:

Reply:
I'm also no Go expert :joy: still learning little by little.

Strictly speaking the controller is mutating the work queue under the hood, updating the controller state in some way, right? But I think ultimately neither options would cause problems in the future

Reply:
That's a valid point. Pointer (no pun intended) it is...